### PR TITLE
Adds Sampler Wrappers for custom samplers in distributed environment

### LINKED ISSUE #9883: Make custom samplers distributed automatically
### Pitch
In DDP mode, PL sets `DistributedSampler` under the hood. However, if you wish to use a custom sampler, then you need to set `Trainer(replace_sampler_ddp=False)` and wrap your custom sampler manually into `DistributedSampler` (https://github.com/PyTorchLightning/pytorch-lightning/issues/5145#issuecomment-745792060)

Instead of having to manually wrap a custom sampler, I think that this can be done under the hood too. For example, https://github.com/pytorch/pytorch/issues/23430#issuecomment-750037457 refers to Catalyst's wrapper for distributing custom samplers. I used it in vanilla PyTorch like this:

```python
from torch.utils.data.distributed import DistributedSampler
from catalyst.data.sampler import DistributedSamplerWrapper

dataset = ...
shuffle = ...
sampler = ...

# If DDP on
if torch.distributed.is_initialized():
    # If using a custom sampler make it distributed
    if sampler is not None:
        sampler = DistributedSamplerWrapper(sampler,
                                            shuffle=shuffle,
                                            num_replicas=communication.get_world_size(),
                                            rank=communication.get_rank())
    # If no custom sampler then just use the DistributedSampler
    else:
        sampler = DistributedSampler(dataset,
                                     shuffle=shuffle,
                                     num_replicas=communication.get_world_size(),
                                     rank=communication.get_rank())

# shuffle shouldn't be specified in DataLoader when using a sampler
shuffle = shuffle if sampler is None else None
dataloader = DataLoader(dataset, sampler=sampler, shuffle=shuffle, ...)
```

cc @borda @awaelchli @rohitgr7 @akihironitta @justusschock @ninginthecloud @tchaton

### LINKED ISSUE #12866: Weighted Random Sampler for ddp
## 🚀 Feature

Creating a version of a weighted random sampler that works efficiently accross multiple GPUs/ nodes.
### Motivation

In my use case . I use a weighted random sampler for Datasets like VQA, and then modify weights across multiple datasets. However, I am a PGR student with limited runtimes available, I switch between debugging locally on single GPUs and production in a HPC cluster. However, "ddp" mode is needed for the HPC, and then my sampler will not work. 

### Pitch

I would like a distributed sampler that behaves the same way as the pytorch WeightedRandomSampler (see PR [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/12413)
### Alternatives

An alternative for my use case might well be to alter "dp" so it can operate in a environment that is not the main thread on a SLURM/SonOfGrid cluster, even when only assigned a single GPU.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @kaushikb11 @awaelchli @akihironitta @rohitgr7

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

Fixes #9883
Closes #12866, #12413

Adds a `DistributedSamplerWrapper` and `UnrepeatedDistributedSamplerWrapper` in order to wrap custom samplers in distributed environment.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @rohitgr7 @akihironitta @justusschock @ninginthecloud @otaj

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: docs/source/common/trainer.rst [hunk 1]
```diff
@@ -1330,6 +1330,7 @@ replace_sampler_ddp
 Enables auto adding of :class:`~torch.utils.data.distributed.DistributedSampler`. In PyTorch, you must use it in
 distributed settings such as TPUs or multi-node. The sampler makes sure each GPU sees the appropriate part of your data.
 By default it will add ``shuffle=True`` for train sampler and ``shuffle=False`` for val/test sampler.
+If you already use a custom sampler, Lightning will wrap it in a way that it samples from your sampler in a distributed manner.
 If you want to customize it, you can set ``replace_sampler_ddp=False`` and add your own distributed sampler.
 If ``replace_sampler_ddp=True`` and a distributed sampler was already added,
 Lightning will not replace the existing one.
```
[the remark this answers is not in the document]
Comment (otaj) [reply docs/source/common/trainer.rst:1336]: I would expect so. I have not touched the "do not replace sampler, if distributed already" codepath and since `DistributedSamplerWrapper` is a subclass of `DistributedSampler`, it should work the same

File: pytorch_lightning/overrides/distributed.py [hunk 2]
```diff
@@ -104,6 +105,88 @@ def __iter__(self) -> Iterator[List[int]]:
...
+
+    Args:
+        sampler: The sampler.
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable]) -> None:
+        if not isinstance(sampler, Sized):
+            raise TypeError("The given sampler must implement the `__len__` method.")
+        if len(sampler) == float("inf"):
+            raise TypeError("The given sampler must be bounded.")
```
[the remark this answers is not in the document]
Comment (otaj) [reply pytorch_lightning/overrides/distributed.py]: I changed the error messages now, what do you think?

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -51,6 +51,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added class name prefix to metrics logged by `DeviceStatsMonitor` ([#12228](https://github.com/PyTorchLightning/pytorch-lightning/pull/12228))
 
 
+- Automatically wrap custom samplers under a distributed environment by using Catalyst's `DistributedSamplerWrapper` ([#12959](https://github.com/PyTorchLightning/pytorch-lightning/pull/12959))
```
Comment (Borda) [inline CHANGELOG.md:54]: Are we setting as a dependency or what is the relation?

File: pytorch_lightning/overrides/distributed.py [hunk 4]
```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
         return iter(indices)
 
 
+class _DatasetFromSampler(Dataset):
+    """Dataset to create indexes from `Sampler`.
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/dataset.py#L6
```
Comment (Borda) [inline pytorch_lightning/overrides/distributed.py:112]: have we clarified the ownership, how significant is the adaptation?
cc: @carmocca

**Review by Borda — CHANGES_REQUESTED** [review]
could we clarify how close are we to the original code?

File: CHANGELOG.md (hunk 3 above)
Comment (otaj) [reply CHANGELOG.md:54]: No, we are not. We adapted two classes from their codebase into ours.

File: pytorch_lightning/overrides/distributed.py [hunk 5]
```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
...
+    def __getitem__(self, index: int) -> Any:
+        """Gets element of the dataset.
+
+        Args:
+            index: index of the element in the dataset
+        Returns:
+            Single element by index
+        """
+        if self._sampler_list is None:
+            self._sampler_list = list(self._sampler)
```
Comment (rohitgr7) [inline pytorch_lightning/overrides/distributed.py:150]: why not initialize it inside init?

File: pytorch_lightning/overrides/distributed.py [hunk 6]
```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
...
+    indices: Tuple = ()
+    if num_indices:
+        indices = itemgetter(*indexes_of_indexes)(dataset)
+        if num_indices == 1:
+            indices = (indices,)  # avoid splitting
+    return iter(indices)
+
+
+class DistributedSamplerWrapper(DistributedSampler):
+    """Wrapper over `Sampler` for distributed training. Allows you to use any sampler in distributed mode.
```
Comment (rohitgr7) [inline pytorch_lightning/overrides/distributed.py:173]:
```suggestion
    """Wrapper over ``Sampler`` for distributed training. Allows you to use any sampler in distributed mode.
```

File: pytorch_lightning/overrides/distributed.py [hunk 7]
```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
...
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/sampler.py#L499
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any) -> None:
+        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
+        self.sampler = sampler
+
+    def __iter__(self) -> Iterator:
+        self.dataset = _DatasetFromSampler(self.sampler)
```
Comment (rohitgr7) [inline pytorch_lightning/overrides/distributed.py:190]:
```suggestion
        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
        self.sampler = sampler

    def __iter__(self) -> Iterator:
        self.dataset = _DatasetFromSampler(self.sampler)
```
just wondering, will they be two different instances: `_DatasetFromSampler(self.sampler)`
why is `self.dataset` required to be instantiated at each iter?

**Comment by Borda** [comment]
> > could we clarify how close are we to the original code?
> 
> @Borda Where would you like us to clarify more? In documentation, in docstrings, in changelog?

probably in the same comment, you mention that it was inspired from...

File: pytorch_lightning/overrides/distributed.py (hunk 5 above)
Comment (otaj) [reply pytorch_lightning/overrides/distributed.py:150]: Because it might be an expensive operation. And, if you only want `__len__`, it is not necessary to have this.

File: pytorch_lightning/overrides/distributed.py (hunk 7 above)
Comment (otaj) [reply pytorch_lightning/overrides/distributed.py:190]: Yes, they will be two different instances. We don't really care about the instance in the `__init__`, since in `__init__` we only care about `__len__`.

But we want to create a fresh iterator everytime we restart the iterator, since the values can change depending on things like seed, etc

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py [hunk 8]
```diff
@@ -104,6 +106,103 @@ def __iter__(self) -> Iterator[List[int]]:
...
+    """Wrapper over ``Sampler`` for distributed training. Allows you to use any sampler in distributed mode.
+
+    It is especially useful in conjunction with :class:`torch.nn.parallel.DistributedDataParallel`.
+    In such case, each process can pass a ``DistributedSamplerWrapper`` instance as a ``DataLoader``
+    sampler, and load a subset of subsampled data of the original dataset that is exclusive to it.
+
+    .. note::
+        Sampler is assumed to be of constant size.
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/sampler.py#L499
```
Comment (carmocca) [inline pytorch_lightning/overrides/distributed.py:182]: Hi @Scitator!

In this PR, we adapted the `DistributedSamplerWrapper` implementation that Catalyst provides after a feature request from the community (see the linked issue).

We were wondering if you had any issues with this proposed piece of code, and whether you accept that we merge it with this reference to the original.

Thank you for your time!

File: src/pytorch_lightning/overrides/distributed.py [hunk 9]
```diff
@@ -104,6 +105,95 @@ def __iter__(self) -> Iterator[List[int]]:
...
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/sampler.py#L499
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any) -> None:
+        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
+        self.sampler = sampler
+
+    def __iter__(self) -> Iterator:
+        self.dataset.reset()
+        return iter(self.dataset[index] for index in super().__iter__())
```
Comment (carmocca) [inline src/pytorch_lightning/overrides/distributed.py:182]: (Not sure) Could this not re-create the iter? Since you already get a generator which I assume is a freshly created iter

```suggestion
        return (self.dataset[index] for index in super().__iter__())
```

File: tests/tests_pytorch/trainer/connectors/test_data_connector.py [hunk 10]
```diff
@@ -271,12 +274,24 @@ class CustomDummyObj:
...
+
+        def __iter__(self):
+            return iter(range(len(self)))
 
     # Should raise an error if existing sampler is being replaced
     dataloader = CustomDataLoader(dataset, sampler=CustomSampler(dataset))
-    with pytest.raises(MisconfigurationException, match="will be replaced by `DistributedSampler`"):
-        trainer._data_connector._prepare_dataloader(dataloader, shuffle=True)
+    result = trainer._data_connector._prepare_dataloader(dataloader)
+    result_dataset = list(iter(result))
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_data_connector.py:290]: I don't think this is requried

```suggestion
    result_dataset = list(result)
```

File: src/pytorch_lightning/overrides/distributed.py (hunk 9 above)
Comment (otaj) [reply src/pytorch_lightning/overrides/distributed.py]: You're absolutely right

File: tests/tests_pytorch/trainer/connectors/test_data_connector.py (hunk 10 above)
Comment (otaj) [reply tests/tests_pytorch/trainer/connectors/test_data_connector.py]: And here as well :)

File: src/pytorch_lightning/overrides/distributed.py [hunk 11]
```diff
@@ -104,6 +105,95 @@ def __iter__(self) -> Iterator[List[int]]:
...
+class UnrepeatedDistributedSamplerWrapper(UnrepeatedDistributedSampler):
+    """Equivalent class to ``DistributedSamplerWrapper`` but for the ``UnrepeatedDistributedSampler``."""
+
+    def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any) -> None:
+        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
+        self.sampler = sampler
+
+    def __iter__(self) -> Iterator:
+        self.dataset.reset()
+        return iter(self.dataset[index] for index in super().__iter__())
```
Comment (carmocca) [inline src/pytorch_lightning/overrides/distributed.py:194]: Forgot this one :)

File: src/pytorch_lightning/overrides/distributed.py [hunk 12]
```diff
@@ -104,6 +105,95 @@ def __iter__(self) -> Iterator[List[int]]:
...
+
+    .. note::
+        Sampler is assumed to be of constant size.
+
+    Adapted from https://github.com/catalyst-team/catalyst/blob/v22.04/catalyst/data/sampler.py#L499
+    """
+
+    def __init__(self, sampler: Union[Sampler, Iterable], *args: Any, **kwargs: Any) -> None:
+        super().__init__(_DatasetFromSampler(sampler), *args, **kwargs)
+        self.sampler = sampler
```
[the remark this answers is not in the document]
Comment (otaj) [reply src/pytorch_lightning/overrides/distributed.py]: Done

File: pytorch_lightning/overrides/distributed.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/overrides/distributed.py]: Hello again! We decided to go with our own implementation in the end.

File: CHANGELOG.md (hunk 3 above)
Comment (carmocca) [inline CHANGELOG.md:54]: Update this too

File: src/pytorch_lightning/overrides/distributed.py (hunk 9 above)
Comment (justusschock) [reply src/pytorch_lightning/overrides/distributed.py]: But this way you do create a tuple out of that and the signature says it needs to return an Iterator, so either we actually do need an `iter()` call or we need to adapt the signature...

Comment (otaj) [reply src/pytorch_lightning/overrides/distributed.py]: I am fairly certain this is not a tuple but generator expression, which is an iterator by itself https://docs.python.org/3/glossary.html#term-generator-expression

File: CHANGELOG.md (hunk 3 above)
Comment (otaj) [reply CHANGELOG.md]: Done

**Comment by otaj** [comment]
At this moment, I believe the PR is ready to go. We decided to move from Catalyst's implementation.

Currently, there is `_DatasetSamplerWrapper` which whole purpose is to wrap `Sampler` into a class with `Dataset`-like features (i.e. inheritance from `Dataset` and meaningful methods `__getitem__` and `__len__`) and `DistributedSamplerWrapper`, which is very much just `DistributedSampler` which takes as a first argument `Sampler` instead of `Dataset`. Under the hood, we create `_DatasetSamplerWrapper` from the `Sampler` and just call the machinery of `DistributedSampler`

@carmocca , can you merge please if the final pass of review still looks good to you? Thanks!
